File: corelib/enumerable.rb
module Enumerable
  def all?(&block)
    %x{
      var result = true;

      if (block !== nil) {
        self.$each._p = function() {
          var value = $opal.$yieldX(block, arguments);

          if (value === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          if (#{Opal.falsy?(`value`)}) {
            result = false;
            return $breaker;
          }
        }
      }
      else {
        self.$each._p = function(obj) {
          if (arguments.length == 1 && #{Opal.falsy?(`obj`)}) {
            result = false;
            return $breaker;
          }
        }
      }

      self.$each();

      return result;
    }
  end

  def any?(&block)
    %x{
      var result = false;

      if (block !== nil) {
        self.$each._p = function() {
          var value = $opal.$yieldX(block, arguments);

          if (value === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          if (#{Opal.truthy?(`value`)}) {
            result = true;
            return $breaker;
          }
        };
      }
      else {
        self.$each._p = function(obj) {
          if (arguments.length != 1 || #{Opal.truthy?(`obj`)}) {
            result = true;
            return $breaker;
          }
        }
      }

      self.$each();

      return result;
    }
  end

  def collect(&block)
    return enum_for :collect unless block_given?

    %x{
      var result = [];

      self.$each._p = function() {
        var value = $opal.$yieldX(block, arguments);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        result.push(value);
      };

      self.$each();

      return result;
    }
  end

  def count(object = undefined, &block)
    %x{
      var result = 0;

      if (object != null) {
        block = function() {
          return #{Opal.destructure(`arguments`) == `object`};
        };
      }
      else if (block === nil) {
        block = function() { return true; };
      }

      self.$each._p = function() {
        var value = $opal.$yieldX(block, arguments);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        if (#{Opal.truthy?(`value`)}) {
          result++;
        }
      }

      self.$each();

      return result;
    }
  end

  def detect(ifnone = undefined, &block)
    return enum_for :detect, ifnone unless block_given?

    %x{
      var result = undefined;

      self.$each._p = function() {
        var params = #{Opal.destructure(`arguments`)},
            value  = $opal.$yield1(block, params);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        if (#{Opal.truthy?(`value`)}) {
          result = params;
          return $breaker;
        }
      };

      self.$each();

      if (result === undefined && ifnone !== undefined) {
        if (typeof(ifnone) === 'function') {
          result = ifnone();
        }
        else {
          result = ifnone;
        }
      }

      return result === undefined ? nil : result;
    }
  end

  def drop(number)
    number = Opal.coerce_to number, Integer, :to_int

    if `number < 0`
      raise ArgumentError, "attempt to drop negative size"
    end

    %x{
      var result  = [],
          current = 0;

      self.$each._p = function() {
        if (number < current) {
          result.push(#{Opal.destructure(`arguments`)});
        }

        current++;
      };

      self.$each()

      return result;
    }
  end

  def drop_while(&block)
    return enum_for :drop_while unless block_given?

    %x{
      var result = [];

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = $opal.$yield1(block, param);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        if (#{Opal.truthy?(`value`)}) {
          return;
        }

        result.push(param);
      };

      self.$each();

      return result;
    }
  end

  def each_slice(n, &block)
    n = Opal.coerce_to n, Integer, :to_int

    return enum_for :each_slice, n unless block_given?

    %x{
      var result,
          slice = []

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)};

        slice.push(param);

        if (slice.length === n) {
          if (block(slice) === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          slice = [];
        }
      };

      self.$each();

      if (result !== undefined) {
        return result;
      }

      // our "last" group, if smaller than n then won't have been yielded
      if (slice.length > 0) {
        if (block(slice) === $breaker) {
          return $breaker.$v;
        }
      }
    }

    nil
  end

  def each_with_index(&block)
    return enum_for :each_with_index unless block_given?

    %x{
      var result,
          index = 0;

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = block(param, index);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        index++;
      };

      self.$each();

      if (result !== undefined) {
        return result;
      }
    }

    nil
  end

  def each_with_object(object, &block)
    return enum_for :each_with_object, object unless block_given?

    %x{
      var result;

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = block(param, object);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }
      };

      self.$each();

      if (result !== undefined) {
        return result;
      }
    }

    object
  end

  def entries
    %x{
      var result = [];

      self.$each._p = function() {
        result.push(#{Opal.destructure(`arguments`)});
      };

      self.$each();

      return result;
    }
  end

  alias find detect

  def find_all(&block)
    return enum_for :find_all unless block_given?

    %x{
      var result = [];

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = $opal.$yield1(block, param);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        if (#{Opal.truthy?(`value`)}) {
          result.push(param);
        }
      };

      self.$each();

      return result;
    }
  end

  def find_index(object = undefined, &block)
    return enum_for :find_index if `object === undefined && block === nil`

    %x{
      var result = nil,
          index  = 0;

      if (object != null) {
        self.$each._p = function() {
          var param = #{Opal.destructure(`arguments`)};

          if (#{`param` == `object`}) {
            result = index;
            return $breaker;
          }

          index += 1;
        };
      }
      else if (block !== nil) {
        self.$each._p = function() {
          var value = $opal.$yieldX(block, arguments);

          if (value === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          if (#{Opal.truthy?(`value`)}) {
            result = index;
            return $breaker;
          }

          index += 1;
        };
      }

      self.$each();

      return result;
    }
  end

  def first(number = undefined)
    %x{
      if (number == null) {
        var result = nil;

        self.$each._p = function() {
          result = #{Opal.destructure(`arguments`)};
          return $breaker;
        };
      }
      else {
        var current = 0,
            result  = [],
            number  = #{Opal.coerce_to number, Integer, :to_int};

        self.$each._p = function() {
          if (number <= current) {
            return $breaker;
          }

          result.push(#{Opal.destructure(`arguments`)});

          current++;
        };
      }

      self.$each();

      return result;
    }
  end

  def grep(pattern, &block)
    %x{
      var result = [],
          proc;

      if (block !== nil) {
        proc = function() {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          var value = #{pattern === `param`};

          if (value !== false && value !== nil) {
            if ((value = block(param)) === $breaker) {
              return $breaker.$v;
            }

            result.push(value);
          }
        };
      }
      else {
        proc = function() {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          var value = #{pattern === `param`};

          if (value !== false && value !== nil) {
            result.push(param);
          }
        };
      }

      #{self}.$each._p = proc;
      #{self}.$each();

      return result;
    }
  end

  def group_by(&block)
    return enum_for :group_by unless block_given?

    hash = Hash.new { |h, k| h[k] = [] }

    %x{
      var result;

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = $opal.$yield1(block, param);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        #{hash[`value`] << `param`};
      }

      self.$each();

      if (result !== undefined) {
        return result;
      }
    }

    hash
  end

  def include?(obj)
    any? { |v| v == obj }
  end

  def inject(object = undefined, &block)
    %x{
      var result = object;

      self.$each._p = function() {
        var value = #{Opal.destructure(`arguments`)};

        if (result === undefined) {
          result = value;
          return;
        }

        value = $opal.$yieldX(block, [result, value]);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        result = value;
      };

      self.$each();

      return result;
    }
  end

  alias map collect

  def max(&block)
    %x{
      var proc, result;
      var arg_error = false;

      if (block !== nil) {
        proc = function() {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          if (result == undefined) {
            result = param;
          }
          else if ((value = block(param, result)) === $breaker) {
            result = $breaker.$v;

            return $breaker;
          }
          else {
            if (value > 0) {
              result = param;
            }

            $breaker.$v = nil;
          }
        }
      }
      else {
        proc = function() {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          var modules = param.$class().__inc__;

          if (modules == undefined || modules.length == 0 || modules.indexOf(Opal.Comparable) == -1) {
            arg_error = true;

            return $breaker;
          }

          if (result == undefined || #{`param` > `result`}) {
            result = param;
          }
        }
      }

      #{self}.$each._p = proc;
      #{self}.$each();

      if (arg_error) {
        #{raise ArgumentError, "Array#max"};
      }

      return (result == undefined ? nil : result);
    }
  end

  def min(&block)
    %x{
      var proc,
          result,
          arg_error = false;

      if (block !== nil) {
        proc = function() {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          if (result == undefined) {
            result = param;
          }
          else if ((value = block(param, result)) === $breaker) {
            result = $breaker.$v;

            return $breaker;
          }
          else {
            if (value < 0) {
              result = param;
            }

            $breaker.$v = nil;
          }
        }
      }
      else {
        proc = function(obj) {
          var param = arguments.length == 1 ?
            arguments[0] : $slice.call(arguments);

          var modules = param.$class().__inc__;

          if (modules == undefined || modules.length == 0 || modules.indexOf(Opal.Comparable) == -1) {
            arg_error = true;

            return $breaker;
          }

          if (result == undefined || #{`param` < `result`}) {
            result = param;
          }
        }
      }

      #{self}.$each._p = proc;
      #{self}.$each();

      if (arg_error) {
        #{raise ArgumentError, "Array#min"};
      }

      return result == undefined ? nil : result;
    }
  end

  alias member? include?

  def none?(&block)
    %x{
      var result = true;

      if (block !== nil) {
        self.$each._p = function() {
          var value = $opal.$yieldX(block, arguments);

          if (value === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          if (#{Opal.truthy?(`value`)}) {
            result = false;
            return $breaker;
          }
        }
      }
      else {
        self.$each._p = function() {
          var value = #{Opal.destructure(`arguments`)};

          if (#{Opal.truthy?(`value`)}) {
            result = false;
            return $breaker;
          }
        };
      }

      self.$each();

      return result;
    }
  end

  def one?(&block)
    %x{
      var result = false;

      if (block !== nil) {
        self.$each._p = function() {
          var value = $opal.$yieldX(block, arguments);

          if (value === $breaker) {
            result = $breaker.$v;
            return $breaker;
          }

          if (#{Opal.truthy?(`value`)}) {
            if (result === true) {
              result = false;
              return $breaker;
            }

            result = true;
          }
        }
      }
      else {
        self.$each._p = function() {
          var value = #{Opal.destructure(`arguments`)};

          if (#{Opal.truthy?(`value`)}) {
            if (result === true) {
              result = false;
              return $breaker;
            }

            result = true;
          }
        }
      }

      self.$each();

      return result;
    }
  end

  def sort_by(&block)
    return enum_for :sort_by unless block_given?

    map {
      arg = Opal.destructure(`arguments`)

      [block.call(arg), arg]
    }.sort.map { |arg| `arg[1]` }
  end

  alias select find_all

  alias reduce inject

  def take(num)
    first(num)
  end

  def take_while(&block)
    return enum_for :take_while unless block

    %x{
      var result = [];

      self.$each._p = function() {
        var param = #{Opal.destructure(`arguments`)},
            value = $opal.$yield1(block, param);

        if (value === $breaker) {
          result = $breaker.$v;
          return $breaker;
        }

        if (#{Opal.falsy?(`value`)}) {
          return $breaker;
        }

        result.push(param);
      };

      self.$each();

      return result;
    }
  end

  alias to_a entries
end

File: corelib/enumerable.js
var RubyEnumerable;

function enum_all() {
  var result = true, iterator = enum_all.proc, func, proc;

  if (iterator) {
    var context = iterator.$S, val;
    enum_all.proc = 0;

    proc = function(e) {
      if ((val = iterator.call(context, e)) === breaker) return breaker.$v;
      if (val === false || val === nil) {
        result = false;
        breaker.$v = nil;
        return breaker;
      }
    };
  }
  else {
    proc = function(e) {
      if (e === false || e === nil) {
        result = false;
        breaker.$v = nil;
        return breaker;
      }
    };
  }

  func = this.m$each;
  func.proc = proc;
  func.call(this);

  return result;
}

function enum_any() {
  var result = false, iterator = enum_any.proc, func, proc;

  if (iterator) {
    var context = iterator.$S, val;
    enum_any.proc = 0;

    proc = function(e) {
      if ((val = iterator.call(context, e)) === breaker) return breaker.$v;
      if (val !== false && val !== nil) {
        result = true;
        breaker.$v = nil;
        return breaker;
      }
    };
  }
  else {
    proc = function(e) {
      if (e !== false && e !== nil) {
        result = true;
        breaker.$v = nil;
        return breaker;
      }
    };
  }

  func = this.m$each;
  func.proc = proc;
  func.call(this);

  return result;
}

function enum_collect() {
  var self = this, iterator = enum_collect.proc;
  if (!iterator) return self.m$enum_for("collect");

  var context = iterator.$S, result = [], val, e;
  enum_collect.proc = 0;

  var proc = function() {
    e = ArraySlice.call(arguments);
    if ((val = iterator.apply(context, e)) === breaker) return breaker.$v;
    result.push(val);
  };

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return result;
}

function enum_count(object) {
  var self = this, iterator = enum_count.proc, result = 0, val;

  if (iterator) {
    var context = iterator.$S;
    enum_count.proc = 0;
  }
  else if (object === undefined) {
    iterator = function() { return true; };
  }
  else {
    iterator = function(e) { return e.m$eq$(object); };
  }

  var proc = function(obj) {
    if ((val = iterator.call(context, obj)) === breaker) return breaker.$v;
    if (val !== false && val !== nil) result++;
  };

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return result;
}

function enum_detect(ifnone) {
  var self = this, iterator = enum_detect.proc;
  if (!iterator) return self.m$enum_for("detect");

  var result = nil, context = iterator.$S, val;
  enum_detect.proc = 0;

  var proc = function(e) {
    if ((val = iterator.call(context, e)) === breaker) return breaker.$v;
    if (val !== false && val !== nil) {
      result = e;
      breaker.$v = nil
      return breaker;
    }
  };

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  if (result !== nil) return result;
  if (typeof ifnone === 'function') return ifnone.m$call();
  return ifnone === undefined ? nil : ifnone;
}

function enum_drop(number) {
  var self = this, result = [], current = 0;
  var proc = function(e) {
    if (number < current) result.push(e);
    current++;
  };

  return result;
}

function enum_drop_while() {
  var self = this, iterator = enum_drop_while.proc;
  if (!iterator) return self.m$enum_for("drop_while");

  var result = [], context = iterator.$S, val;
  enum_drop_while.proc = 0;

  var proc = function(e) {
    if ((val = iterator.call(context, e)) === breaker) return breaker.$v;
    if (val !== false && val !== nil) result.push(e);
    else return breaker;
  };

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return result;
}

function enum_each_with_index() {
  var self = this, iterator = enum_each_with_index.proc;
  if (!iterator) return self.m$enum_for("each_with_index");

  var index = 0, context = iterator.$S, val;

  var proc = function(e) {
    if ((val = iterator.call(context, e, index)) === breaker) return breaker.$v;
    index++;
  };

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return nil;
}

function enum_entries() {
  var self = this, result = [];
  var proc = function(e) { return result.push(e); };
  var func = self.m$each;
  func.proc = proc;
  func.call(self);
  return result;
}

function enum_find_index(object) {
  var self = this, iterator = enum_find_index.proc, val, result = nil;
  if (object === undefined && !iterator) return self.$enum_for("find_index");

  if (object !== undefined) iterator = function(e) { return e.m$eq$(object); };
  else var context = iterator.$S;

  enum_find_index.proc = 0;

  var proc = function(obj, index) {
    if ((val = iterator.call(context, obj)) === breaker) return breaker.$v;
    if (val !== false && val !==nil) {
      result = obj;
      breaker.$v = index;
      return breaker;
    }
  };

  var func = self.m$each_with_index;
  func.proc = proc;
  func.call(self);

  return result;
}

function enum_first(number) {
  var result = [], proc;

  if (number === undefined) {
    proc = function(e) {
      result = e;
      return breaker;
    };
  }
  else {
    var current = 0;
    proc = function(e) {
      if (number < current) return breaker;
      result.push(e);
      current++;
    };
  }

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return result;
}

function enum_grep(pattern) {
  var self = this, ary = [], iterator = enum_grep.proc;

  if (iterator) {
    var context = iterator.$S, val;
    enum_grep.proc = 0;

    var proc = function(e) {
      if (val = pattern.m$eqq$(e), val !== false && val !== nil) {
        if ((val === iterator.call(context, e)) === breaker) return breaker.$v;
        ary.push(e);
      }
    };
  }
  else {
    var proc = function(e) {
      if (val === pattern.m$eqq$(e), val !== false && val !== nil) {
        ary.push(e);
      }
    };
  }

  var func = self.m$each;
  func.proc = proc;
  func.call(self);

  return ary;
}

function init_enumerable() {
  RubyEnumerable = define_module(rb_cObject, 'Enumerable');
  define_method(RubyEnumerable, 'm$all$p', enum_all);
  define_method(RubyEnumerable, 'm$any$p', enum_any);
  define_method(RubyEnumerable, 'm$collect', enum_collect);
  define_method(RubyEnumerable, 'm$count', enum_count);
  define_method(RubyEnumerable, 'm$detect', enum_detect);
  define_method(RubyEnumerable, 'm$find', enum_detect);
  define_method(RubyEnumerable, 'm$drop', enum_drop);
  define_method(RubyEnumerable, 'm$drop_while', enum_drop_while);
  define_method(RubyEnumerable, 'm$each_with_index', enum_each_with_index);
  define_method(RubyEnumerable, 'm$entries', enum_entries);
  define_method(RubyEnumerable, 'm$find_index', enum_find_index);
  define_method(RubyEnumerable, 'm$first', enum_first);
  define_method(RubyEnumerable, 'm$grep', enum_grep);

  define_method(RubyEnumerable, 'm$to_a', enum_entries);
}
